### PR TITLE
vmauth: reader pool to reduce gc & mem alloc

### PR DESCRIPTION
**Pull request by shichanglin5**
goto https://github.com/VictoriaMetrics/VictoriaMetrics/issues/6445

### CODE REVIEW

**Comment by shichanglin5** [comment]
After the change:
1. If the request body is not completely cached to buf when the request failed due to write broken before, it cannot be retried. Now it is OK (read buf + stream)
2. reduce gc and memory allocation, observation and monitoring can see a significant improvement
image

**Review by f41gh7 — CHANGES_REQUESTED** [review]
Thanks for the pull request!

 Could you please rebase it to the master branch and target PR also for master branch.


 I have few minor comments, which should simplify code reading.

 Feel free to address this changes or we can create separate PR based on your changes with needed fixes

File: app/vmauth/main.go [hunk 1]
```diff
@@ -200,10 +202,25 @@ func processRequest(w http.ResponseWriter, r *http.Request, ui *UserInfo) {
 		up, hc = ui.DefaultURL, ui.HeadersConf
 		isDefault = true
 	}
-	maxAttempts := up.getBackendsCount()
-	r.Body = &readTrackingBody{
-		r: r.Body,
+	if r.ContentLength <= int64(maxRequestBodySizeToRetry.IntN()) {
+		bufReader := bufPool.Get()
```
Comment (f41gh7) [inline app/vmauth/main.go:206]: it's better to create helper get and put functions like: 

```go
var readTrackingBodyPool sync.Pool

func getReadTrackingBody(origin io.ReadCloser, b int) *readTrackingBody {
	bufSize := 1024
	if b > 0 {
		bufSize = b
	}
	v := readTrackingBodyPool.Get()
	if v == nil {
		rtb := &readTrackingBody{
			r:   origin,
			buf: make([]byte, 0, bufSize),
		}
		return rtb
	}
	rtb := v.(*readTrackingBody)
	rtb.r = origin
	rtb.buf = bytesutil.ResizeNoCopyMayOverallocate(rtb.buf, b)[:0]

	return rtb
}

func putReadTrackingBody(rtb *readTrackingBody) {
	rtb.reset()
	readTrackingBodyPool.Put(rtb)
}
```

Comment (f41gh7) [reply app/vmauth/main.go:206]: it simplifies pool usage and always reset instance properties at put

File: app/vmauth/main.go [hunk 2]
```diff
@@ -503,60 +520,70 @@ type readTrackingBody struct {
...
-		rtb.cannotRetry = true
-		return n, err
+	if n > 0 {
+		rtb.offset += n
+		rtb.buf = append(rtb.buf, p[:n]...)
 	}
-	rtb.buf = append(rtb.buf, p[:n]...)
 	if err == io.EOF {
 		rtb.bufComplete = true
+	} else if err != nil {
```
Comment (f41gh7) [inline app/vmauth/main.go:574]: From reader point of view, it's better to check err != nil and perform fast path on specific error. 

 Like:
```go

if err != nil {
		if err == io.EOF {
			rtb.bufComplete = true
			return n, err
		}
		rtb.cannotRetry = true
	}
```

File: app/vmauth/main.go [hunk 3]
```diff
@@ -503,60 +520,70 @@ type readTrackingBody struct {
...
+		if rtb.bufComplete {
+			if rtb.offset == len(rtb.buf) {
+				return nb, io.EOF
+			}
+			return nb, nil
+		}
+		if nb < len(p) {
+			nr, err := rtb.readFromStream(p[nb:])
+			if err != nil {
+				return 0, err
```
Comment (f41gh7) [inline app/vmauth/main.go:545]: I'd suggest to always return number of bytes read:

```go
nr, err := rtb.readFromStream(p[nb:])
return nb + nr, err
```

**Comment by shichanglin5** [comment]
a new pr based on branch master: https://github.com/VictoriaMetrics/VictoriaMetrics/pull/6533

**Pull request closed without merging**